extras/indexing/src/main/java/org/apache/rya/indexing/entity/EntityIndexSetProvider.java

```diff
@@ -0,0 +1,136 @@
+     * @param entityStorage - The mechanism for access to the various {@link Entity}s. (not null)
+     */
+    public EntityIndexSetProvider(final TypeStorage typeStorage, final EntityStorage entityStorage) {
+        this.typeStorage = requireNonNull(typeStorage);
+        this.entityStorage = requireNonNull(entityStorage);
+    }
+
+    private Var getTypeSubject(final Type type) {
+        //we just need the first pattern since all the patterns in this type map are the same subject
+        final StatementPattern pattern = typeMap.get(type).iterator().next();
+        return pattern.getSubjectVar();
+    }
+
+    private RyaURI getPredURI(final StatementPattern pattern) {
+        final Var pred = pattern.getPredicateVar();
+        return new RyaURI(pred.getValue().stringValue());
+    }
+
+    @Override
+    public List<EntityQueryNode> getExternalSets(final QuerySegment<EntityQueryNode> node) {
+        typeMap = HashMultimap.create();
+        subjectTypeMap = new HashMap<>();
+
+        //discover entities
+        final List<StatementPattern> unused = new ArrayList<>();
+        for (final QueryModelNode pattern : node.getOrderedNodes()) {
+            if(pattern instanceof StatementPattern) {
+                discoverEntities((StatementPattern) pattern, unused);
+            }
+        }
+
+        final List<EntityQueryNode> nodes = new ArrayList<>();
+        for(final Type type : typeMap.keySet()) {
+            //replace all nodes in the tupleExpr of the collection of statement patterns with this node.
+            final EntityQueryNode entity = new EntityQueryNode(type, typeMap.get(type), entityStorage);
+            nodes.add(entity);
+        }
+        return nodes;
+    }
+
+    private void discoverEntities(final StatementPattern pattern, final List<StatementPattern> unmatched) {
+        final Var subj = pattern.getSubjectVar();
+        final String subjStr = subj.getName();
+        final RyaURI predURI = getPredURI(pattern);
+        //check to see if current node is type
+        if(new URIImpl(predURI.getData()).equals(RDF.TYPE)) {
+            final Var obj = pattern.getObjectVar();
+            final RyaURI objURI = new RyaURI(obj.getValue().stringValue());
+            try {
+                final Optional<Type> optType = typeStorage.get(objURI);
+                //if is type, fetch type add to subject -> type map
+                if(optType.isPresent()) {
+                    final Type type = optType.get();
+                    typeMap.put(type, pattern);
+                    subjectTypeMap.put(subjStr, type);
+                    //check unmatched properties, add matches
+                    for(final StatementPattern propertyPattern : unmatched) {
+                        //store sp's as uri?
+                        final RyaURI property = getPredURI(propertyPattern);
+                        final Var typeSubVar = getTypeSubject(type);
```

Comment:
Why call getTypeSubject here to extract the subject Var from typeMap?  You already have access to subjStr.  Why not just call propertyPattern.getSubjectVar().getName() below and do a String comparison in your if clause?